### PR TITLE
#2921 The Generate Outbound (Sales Order/Distribution) the Weight and…

### PR DESCRIPTION
**Pull request by e-Evolution**
… Volumen show incorrect info https://github.com/adempiere/adempiere/issues/2921

- Add Delivery Rule and Delivery Via
- Add the Attribute Set Instance Id
- Add Calculate Delivery to to multiples Attribute Set Instance Id

### CODE REVIEW

**Comment by marcalwestf** [comment]
Hi @e-Evolution Victor, I tried to test this PR first on Swing, but there is no V class, so I wonder if you forgot to include it:
Selection_983

When trying to open the Smart Browser _Generate Outbound (Sales Order/Distribution Order)_, I get a _Class Not Found_ Exception:
Selection_984

I will continue to test on ZK in the meantime.

**Comment by marcalwestf** [comment]
One remark: when an Order is selected, lines are displayed with a weight and volume calculation which depends on the quantity. **This is OK**:
Selection_986

when the lines are selected, weight and volume are rounded (because the unit is _Each_), **which is correct**.
Selection_987

The only improvement here would be to round also correctly **before** the lines are selected.

**Comment by yamelsenih** [comment]
Hi @marcalwestf @e-Evolution already exist other PR for this and my test was failed, I want to test it before merge in ADempiere

**Comment by marcalwestf** [comment]
... I created a product with an UOM=kg, fractional weight and volume; then completed a Sales Order with a fractional quantity.

The Weight and Volume of the selected line are nevertheless rounded, which I think is an error because volume, as well as weight, should in these cases be fractions.
Selection_988

**Comment by marcalwestf** [comment]
@yamelsenih ok. 
Which PR is about? Thanks.

**Comment by marcalwestf** [comment]
concerning my initial comment that the V class is missing: it was so from the beginning.

**Comment by marcalwestf** [comment]
..waiting for @yamelsenih Yamel to carry out tests before accepting this PR.

**Comment by yamelsenih** [comment]
Testing ti now

**Comment by yamelsenih** [comment]
Testing it

File: org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java
Comment (yamelsenih) [inline org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java:200]: I don't sure if this is correct, maybe is a change for a customer or own change, the Weigh and Volume is showed from Order to select instead of Quantoty to deliver.

I think that this change can be part of a ADempiere Labs, Customizxation for your customer otr a Location

File: org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java
Comment (yamelsenih) [inline org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java:263]: I don't sure if this is correct, maybe is a change for a customer or own change, the Weigh and Volume is showed from Order to select instead of Quantoty to deliver.

I think that this change can be part of a ADempiere Labs, Customizxation for your customer otr a Location

File: org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java
Comment (yamelsenih) [inline org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java:1346]: Please make this code more atomic, the refactory can be change in other pull request as Source code Improve

File: org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java
Comment (yamelsenih) [inline org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java:1363]: Please make this code more atomic, the refactory can be change in other pull request as Source code Improve

File: org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java
Comment (yamelsenih) [inline org.eevolution.warehouse/src/main/java/base/org/spin/form/OutBoundOrder.java:1384]: Please make this code more atomic, the refactory can be change in other pull request as Source code Improve

**Review by yamelsenih — CHANGES_REQUESTED** [review]

**Comment by yamelsenih** [comment]
I am waiting for resolve all suggestions

**Comment by e-Evolution** [comment]
In any case, I am trying to improve this functionality, although its application is really limited and it cannot be used in an environment where you have multiple people preparing shipments since when calculating picking on memory two people may be cause conflict for the same existence, for what this form has a use limited to one person.

That is why the process-based approach is more reliable since it is done through an atomic transaction so you do not have the problem of conflict in the use of existence.

Maybe @yamelsenih  Yamel is right and we should remove it from the core system and only keep it as customization due to its limited scope or the possibility of causing problems in the picking

**Comment by yamelsenih** [comment]
Hi @e-Evolution I use this funciionality for many customers without problems. Your changes only change the current behavior and I think that could can implent your own changes for your customization. The Generate outbound order was created for generate a outbound based on current sales orders or distribution and inventory. The attributes management is nice and I think that can be included but my test with it have a result very distint.

In other hand. The weight and volume is based on order currently and I think that calculate it I not necessary because already exist a calculation in status bar when is selected a record.

Best regards

**Comment by e-Evolution** [comment]
Well, @yamelsenih  say what you want, even using the functionality that you introduced, does not work in 3.9.3 and I put the steps to replicate the problem.

is clear that exists  a concurrency problem and it was demonstrated step by step with a business case, you can close your eyes or tell me I don't understand

always is easier answering anything than using answering with arguments.

Anyway, @marcalwestf Mario will review the case, I prefer that he review and accept or not,  as he took the time to deploy and review with detail,  and it may even be better to disable this functionality, firstly because there is no implementation for swing and because not work.

**Comment by yamelsenih** [comment]
Hi @e-Evolution I don't agree with merge it. If exist a bug with concurrent then the results of bot (before and after) must be the same and the change proposed is not the same. Also add changes like weight and volume distint of original propose of form. I add all comment in the pull request and I am waiting for each correction or reply 

Best regards

**Comment by yamelsenih** [comment]
Other comment: If you want to solve bug then limit your pull request to only solve it instead.

This add more functionality and buggies. Is not atomic.

> Add Delivery Rule and Delivery Via
>Add the Attribute Set Instance Id
>Add Calculate Delivery to to multiples Attribute Set Instance Id

The. What is the real scope of this pr? Make a fix of bug or also is enhance?

Please make other PR or change this for make it more atomic and test with a real case.

Best regards

**Comment by marcalwestf** [comment]
@e-Evolution  @yamelsenih  Victor and Yamel, how do we here go on?

**Comment by yamelsenih** [comment]
I tasted it and not get the same result for query. I can test another time now

**Comment by yamelsenih** [comment]
Testing it:

First comment:
This form is was created for improve the distribution management and exist columns that was designed for show the information more fast to operator. I saw that all columnas was changed and it is not good.
The main interest for a Logistic Manager is know what is the delivery route.
Before change:
Before_1
After change:
After_1

Recomendation: I am agree with add new columns but follow the original desing you can add they after last instead

**Comment by yamelsenih** [comment]
Second comment:
For Attribute set instance I think tah this change is really important and should be added but a only chanche without change all desing. I agree with add it but is innecessary change all form just for add support to attribute set instance.

**Comment by yamelsenih** [comment]
Third comment (Tech comment):
I created the Generate Outbound Order wwith many care of get the maximum speed for operations, Note that not exist Sub-querys and other bad use of SQL. I can see that the weight and volume is added ad Sub-query and it is very bad for performance of query:

```SQL
(SELECT SUM((ol.QtyOrdered - ol.QtyDelivered) * p.Weight)  AS Weight FROM DD_OrderLine ol INNER JOIN M_Product p ON(p.M_Product_ID = ol.M_Product_ID) WHERE ol.DD_Order_ID = ord.DD_Order_ID) AS Weight
```

Exists many ways (not easy but better like) SELECT AS FROM and other way for use aggregate functions instead it

**Comment by yamelsenih** [comment]
Fourth comment:
The additional columns like weigh and volume already exists as part of form. The just get the weith and volume of Sales / Distribution order (This calculation is created with the model of MOrder and MDDistribution) I think that this is very rounded because already exists a business logic for it and is unnecessary make other calculation here. Also the total weigh and total volume can be showed when you select the lines of Orders, see:
Screenshot_20210710_080707
Screenshot_20210710_080725

**Comment by yamelsenih** [comment]
I suggest create a Pull Request for add support to Attribute set instance and other adding the additional columns

**Comment by yamelsenih** [comment]
Hello @e-Evolution can you update this pull request? or create a new instead?

**Comment by e-Evolution** [comment]
Close open other PR

**Pull request closed without merging**
